### ISSUE TITLE
Using cxx in no-std context

### ISSUE THREAD

**Issue by harrier-lcc** [dtolnay/cxx#933]
I am trying to explore the possibility to use cxx in embedded development, for using C/C++ low-level API with Rust. However, I have to use `no_std` in my target platform.

Is there any way to use `cxx::bridge` in `no_std` context? Most embedded platform out there uses C/C++ API so this would help a lot to integrate existing Rust code and libraries.

**Comment by dtolnay** [comment]
As long as you have a liballoc implementation for your platform, I think it would be straightforward for cxx to support. No alloc is also in scope for cxx but would be a more involved change that I probably wouldn't prioritize in the near term.

**Comment by harrier-lcc** [comment]
Currently `link_cplusplus` requires `cc` which requires `std`, which block the usage of `cxx` in `no_std` directly even with liballoc.

Will using `cxxbridge_macro` directly for `cxx::bridge` work if I move the link and build process in non-cargo platform like cmake?

**Comment by aditsachde** [comment]
Were you able to get it to work using `cxxbridge_macro`? I'm running into the same issue as I'd like to use a C++ library from embedded rust via CXX.

**Comment by dtolnay** [comment]
> Currently `link_cplusplus` requires `cc` which requires `std`, which block the usage of `cxx` in `no_std` directly even with liballoc.

This is not the case. Link-cplusplus has only a build dependency on cc, not a runtime dependency. Build dependencies and macros are always compiled for the host, with std, even when building for a no-std target.

**Comment by harrier-lcc** [comment]
> > Currently `link_cplusplus` requires `cc` which requires `std`, which block the usage of `cxx` in `no_std` directly even with liballoc.
> 
> This is not the case. Link-cplusplus has only a build dependency on cc, not a runtime dependency. Build dependencies and macros are always compiled for the host, with std, even when building for a no-std target.

You are right on `cc` is only a build dep (didn't look that carefully enough), but as `link_cplusplus` don't declare `#![no_std]` it makes cxx cant be compiled with no std targets.

**Comment by harrier-lcc** [comment]
the `link_cplusplus` implicit std requirement is resolved thanks to @dtolnay

though in 
https://github.com/dtolnay/cxx/blob/178b0b6b4992d83fc9df443d55ec359ddd4e4932/src/lib.rs#L400
it does `extern crate std;` without conditional compiles which somehow contradicts the `#![no_std]` tags.

**Comment by dtolnay** [comment]
I published 1.0.58 with an attempt at no-std support:

```toml
[dependencies]
cxx = { version = "1.0.58", default-features = false, features = ["alloc"] }
```

Please let me know whether this does what you need.
